zebra-state/src/service/read/tree.rs

```diff
@@ -66,78 +66,46 @@ where
 pub fn sapling_subtrees<C>(
     chain: Option<C>,
     db: &ZebraDb,
-    start_index: NoteCommitmentSubtreeIndex,
-    limit: Option<NoteCommitmentSubtreeIndex>,
+    range: impl std::ops::RangeBounds<NoteCommitmentSubtreeIndex> + Clone,
 ) -> BTreeMap<NoteCommitmentSubtreeIndex, NoteCommitmentSubtreeData<sapling::tree::Node>>
 where
     C: AsRef<Chain>,
 {
-    let mut db_list = db.sapling_subtree_list_by_index_for_rpc(start_index, limit);
+    use std::ops::Bound::*;
 
-    if let Some(limit) = limit {
-        let subtrees_num = u16::try_from(db_list.len())
-            .expect("There can't be more than `u16::MAX` Sapling subtrees.");
-
-        // Return the subtrees if the amount of them reached the given limit.
-        if subtrees_num == limit.0 {
-            return db_list;
-        }
-
-        // If not, make sure the amount is below the limit.
-        debug_assert!(subtrees_num < limit.0);
-    }
-
-    // If there's no chain, then we have the complete list.
-    let Some(chain) = chain else {
-        return db_list;
-    };
-
-    // Unlike the other methods, this returns any trees in the range,
-    // even if there is no tree for start_index.
-    let fork_list = chain.as_ref().sapling_subtrees_in_range(start_index, limit);
-
-    // If there's no subtrees in chain, then we have the complete list.
-    if fork_list.is_empty() {
-        return db_list;
+    let start_index = match range.start_bound().cloned() {
```

Comment:
If the goal is to find the first included index, we're effectively doing something like:

```rust
let start_index = (0..=u16::MAX).iter().find(|index| range.contains(NoteCommitmentSubtreeIndex(index)));
let Some(start_index) = start_index else {
    // Range is mathematically empty
    return BTreeMap::new();
};
```

Ideally writing that could would optimise to start_index, start_index + 1, or 0 based on the bound. But that doesn't seem to happen yet, maybe Rust iterators don't optimise constant iterators with closures well.

You can see the assembly here, it's significantly longer than just matching the bound:
 https://play.rust-lang.org/?version=stable&mode=release&edition=2021